File: leetcode/C++/binary-tree-cameras.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int answer;
    void installCamera(TreeNode *parent, TreeNode *node){
        if(parent) ++parent->val;
        if(node->left) ++node->left->val;
        if(node->right) ++node->right->val;
        ++node->val;
        ++answer;
    }
    void checkCameras(TreeNode *parent, TreeNode *node){
        if(!node) return;
        checkCameras(node,node->left);
        checkCameras(node,node->right);
        if(node->left && node->left->val==0) installCamera(parent,node);
        else if(node->right && node->right->val==0) installCamera(parent,node);
    }
    int minCameraCover(TreeNode* root) {
        answer=0;
        checkCameras(NULL,root);
        return (root->val)? answer : answer+1;
    }
};

File: leetcode/Python/binary-tree-cameras.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.ans=0
    def installCamera(self, parent, node):
        if parent:
            parent.val+=1
        if node.left:
            node.left.val+=1
        if node.right:
            node.right.val+=1
        node.val+=1
        self.ans+=1
    def checkCameras(self, parent, node):
        if not node:
            return
        self.checkCameras(node,node.left)
        self.checkCameras(node,node.right)
        if node.left and node.left.val == 0:
            self.installCamera(parent,node)
        if node.right and node.right.val == 0:
            self.installCamera(parent,node)
    def minCameraCover(self, root: TreeNode) -> int:
        self.checkCameras(None,root)
        if root.val == 0:
            self.ans+=1
        return self.ans
        